File: level2/조이스틱/solution.js
function solution(name) {
  var answer = 0;
  let len = name.length;
  //각각 문자의 최소 상하이동은 모두 더해둔다.
  for (var i = 0; i < len; i++) {
    let cca = name.charCodeAt(i);
    answer +=
      name.charAt(i) > "M"
        ? "Z".charCodeAt() - cca + 1
        : cca - "A".charCodeAt();
  }

  let min = len - 1;
  // A가 포함되지 않으면 왼쪽부터 끝까지 가는게 최소이동
  if (name.indexOf("A") === -1) return answer + min;

  for (let i = 1; i < len; i++) {
    if (name[i] === "A") {
      for (var j = i + 1; j < len; j++) {
        if (name[j] !== "A") {
          break;
        }
      }

      const left = i - 1;
      const right = len - j;
      // 다른 사람 풀이를 참고 했는데 *2가 해깔린다...
      min = Math.min(min, left > right ? left + right * 2 : left * 2 + right);

      i = j;
    }
  }
  return answer + min;
}

console.log(solution("BBBAAABBAABBABB"));
console.log(solution("JEROEN"));
console.log(solution("JAN"));
console.log(solution("BBBBBBBBBABB"));
console.log(solution("AAAAAAAABAAAAAAAA"));

File: level2/조이스틱/solution.rb
def solution(name)
  answer = 0
  idxs = []
  name.chars.each_with_index do |s, i|
    # 각 문자 상하 최소 이동거리 더하기
    answer += s > "M" ? "Z".ord - s.ord + 1 : s.ord - "A".ord
    prev = i == 0 ? "" : name[i - 1]
    # 왼쪽부터 A전까지 이동거리
    idxs << i - 1 if s == "A" && prev != s
    # 오른쪽부터 A전까지 이동거리
    idxs << name.size - i if s != "A" && prev == "A"
  end
  # A로 시작하는 경우 처음에 -1이 들어감
  idxs[0] = 0 if idxs[0] == -1
  # 홀수개 인 경우는 뒤쪽에 A가 붙은 경우
  idxs << 0 if idxs.size.odd?

  # 최소 이동거리 계산
  min = name.size - 1
  0.step(idxs.size - 1, 2) do |i|
    sum = idxs[i] + idxs[i + 1]
    min = sum if min > sum
  end
  return answer + min
end

p solution "JEROEN"
p solution "JAN"
p solution "BAAABBBAAABBAABBABBA"
p solution "BBBBBBBBBABB"
p solution "AAAAAAAABAAAAAAAA"
